### PR TITLE
Param path function

### PR DESCRIPTION
**Pull request by robinsteuteville**

### CODE REVIEW

File: python/altrios/__init__.py [hunk 1]
```diff
@@ -31,3 +31,228 @@ def __array__(self):
...
+
+def history_path_list(self) -> list[str]:
+    """Returns a list of relative paths to all history variables (all variables
+    that contain history as a subpath). 
+    See example usage in param_paths_demo.py."""
+    return [item for item in self.param_path_list() if "history" in item]
+            
+
+# add param_path_list as an attribute for all Rust structs
+setattr(FuelConverter, "param_path_list", param_path_list)
```
Comment (calbaker) [inline python/altrios/__init__.py:137]: @robinsteuteville , per our discussion with Kyle, change this to something like 
```python
setattr(getattr(alt, "FuelConverter"), "param_path_list", param_path_list)
```

File: python/altrios/tests/test_param_paths.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            mock_df['time [s]'],
+            mock_df['Tractive Power [W]'],
+            engine_on=mock_df['engine_on'],
+        )
+        loco_sim = LocomotiveSimulation(
+            loco_unit=Locomotive.default(),
+            power_trace=pt,
+            save_interval=1)
+
+        baseline_variable_paths = ['power_trace.time_seconds', 'power_trace.pwr_watts', 'power_trace.time_hours', 'power_trace.engine_on', 'loco_unit.edrv.save_interval',
```
Comment (calbaker) [inline python/altrios/tests/test_param_paths.py:19]: I think I'd like to see this list in a separate file (not sure where pathwise but; see if you can find a good location) and loaded by the script using e.g. 
```python
with Open(fsim.resources_dir() / "test_assets/baseline_variable_paths.txt"
```
where the file is a plain text file with 1 string per line.

File: python/altrios/__init__.py [hunk 3]
```diff
@@ -31,3 +31,228 @@ def __array__(self):
...
+                         'RailVehicle',
+                         'TrainSimBuilder',
+                         'SpeedLimitTrainSimVec',
+                         'EstTimeNet',
+                         'Pyo3VecWrapper',
+                         'Pyo3Vec2Wrapper',
+                         'Pyo3Vec3Wrapper',
+                         'Pyo3VecBoolWrapper']
+
+def param_path_list(self, path = "", param_path_list = []) -> list[str]:
```
Comment (calbaker) [inline python/altrios/__init__.py:92]: I think it's better to call this `var_path_list` or `variable_path_list`.  I usually use `param` for things that the user specifies.   Please change and propagate if this also makes sense to you.

File: python/altrios/__init__.py (hunk 1 above)
Comment (calbaker) [reply python/altrios/__init__.py:137]: @robinsteuteville , be sure to propagate this to fastsim

File: python/altrios/__init__.py (hunk 3 above)
Comment (calbaker) [reply python/altrios/__init__.py:92]: @robinsteuteville , be sure to propagate this to fastsim

File: python/altrios/demos/param_paths_demo.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# %%
```
Comment (calbaker) [inline python/altrios/demos/param_paths_demo.py:1]: Change to:
```python
Script demonstrating how to use param_path_list() and history_path_list()
demos to find the paths to variables within altrios classes.
```

This is how you put a doc string in at the script/module level.

File: python/altrios/__init__.py [hunk 5]
```diff
@@ -31,3 +31,228 @@ def __array__(self):
...
+    """Returns list of relative paths to all variables and sub-variables within
+    class (relative to the class the method was called on) 
+    See example usage in param_paths_demo.py.
+    Arguments:
+    ----------
+    path : Defaults to empty string. This is mainly used within the method in
+    order to call the method recursively and does not need to be specified by
```
Comment (calbaker) [inline python/altrios/__init__.py:99]: @robinsteuteville 
change  

> method recursively and does not need to be specified by user
to 

> method recursively and should not be specified by user

File: python/altrios/__init__.py [hunk 6]
```diff
@@ -104,15 +104,17 @@ class (relative to the class the method was called on)
...
+    variable_list = [attr for attr in self.__dir__() if not attr.startswith("__")\
+         and not callable(getattr(self,attr))]
     for variable in variable_list:
         if not type(getattr(self,variable)).__name__ in ACCEPTED_RUST_STRUCTS:
             if path == "":
                 variable_path = variable
             else:
                 variable_path = path + "." + variable
             param_path_list.append(variable_path)
-        elif len([attr for attr in getattr(self,variable).__dir__() if not attr.startswith("__") and not callable(getattr(getattr(self,variable),attr))]) == 0:
```
Comment (calbaker) [inline python/altrios/__init__.py]: @robinsteuteville for long lines, you can shorten them by ending with `\`.  Ensure that the `\` is right after the last character of the code on that line and that there is no trailing whitespace.

**Comment by robinsteuteville** [comment]
@calbaker I have updated altrios and fastsim based on your comments, but I found a bug which will need to be fixed before it can be merged.

**Comment by calbaker** [comment]
@robinsteuteville, I found an example of a missing path:
```python
train_sim.loco_con.loco_vec.tolist()[0].edrv.history.pwr_mech_prop_out_watts
```

You can see this by running to this line:
https://github.com/NREL/altrios/blob/98f098d88bf6fd568848806c73816ec7fc050a1d/python/altrios/demos/speed_limit_simple_corr_demo.py#L94

I'm not sure this same edge case will come up in fastsim, but it may be worth looking a bit.  I think I'd propose checking for this type:
```
In [11]: type(train_sim.loco_con.loco_vec)
Out[11]: altrios.Pyo3VecLocoWrapper
```
or checking for the presence if a `tolist` method. 

See if you can figure something out.  I'd also recommend benchmarking the test against something at the `train_sim` level.

**Comment by calbaker** [comment]
@robinsteuteville , I'm also not sure whether to include the `tolist()` method call in the path.  If it's a path for someone to copy/paste, then I'm thinking `tolist` should be retained.

**Comment by robinsteuteville** [comment]
@calbaker `Pyo3VecLocoWrapper` is not appearing in `altrios.__dir__()` (which is how `variable_path_list()` is automatically created as a method for other Rust structs). When I tried to manually add `variable_path_list()` as a method for `Pyo3VecLocoWrapper`, I got the error `AttributeError: partially initialized module 'altrios' has no attribute 'Pyo3VecLocoWrapper' (most likely due to a circular import).`. I'm wondering if this is because `Pyo3VecLocoWrapper` is not actually included in the `altrios-py/src/lib.rs` file (as in, there is no `m.add_class::<Pyo3VecLocoWrapper>()?;`. Wondering if this needs to be added, or if this was left out intentionally?

**Comment by calbaker** [comment]
@robinsteuteville 
>  this is because Pyo3VecLocoWrapper is not actually included in the altrios-py/src/lib.rs file (as in, there is no m.add_class::<Pyo3VecLocoWrapper>()?;. 
This is almost certainly the root cause.  

> Wondering if this needs to be added, or if this was left out intentionally?
Try it and see what happens!

**Comment by SWRIganderson** [comment]
@calbaker This is your long lost friend.  I do not get a nice list when I run the demo.  I have this installed in an anaconda environment running python 3.10.  I pulled commit 922a250.  I saved the output that was generated before the error in the text file below.  The error I get is 

[VariableDemoOutput.txt](https://github.com/user-attachments/files/16199414/VariableDemoOutput.txt)
 
image

I also noticed something odd when I was setting up the new environment.  I usually start from scratch when I'm checking things out for you if it is a new branch to make sure I don't have something off on my machine.  When I did the pip install, I first tried from requirements.txt.  It threw the error below.  We may have a problem with package dependencies in the requirement file that needs to be address.  The pip e dev command worked fine.

image

**Comment by SWRIganderson** [comment]
@calbaker I played around with a few other objects in the demo. I still get the None error for most of the others too.  From what I understand, it will spit out all paths and values.  I could see there being two needs here.  The first is just getting paths so you can figure out what all you have to play with.  This is quite useful.  My crutch is the spyder variable editor for this, but I must first convert to json and load the json in spyder.  Its a bit clunky, but what can you expect from a guy who went to A&M and UT.  I'm a divided house.  

The second need is to spit out paths and data as it is currently doing.  I could see this being quite useful for post processing data.  I did something similar for the taconite work with the JSON Spyder dance.  However, I think I settled on CBOR which is a binary JSON equivalent.  

The question I would pose here.  Are there already tools that would exist for yaml, json, or cbor that would print out all paths in a structure.  Or all paths with values?  I realize that CBOR is not int he main branch right now, but it can be added in about 15 minutes with 3 or 4 lines of code.  I'm good with however you want to approach this.  Just not sure if it would be faster to complete with a different approach or give us more features in the long run with less code maintenance.

**Comment by SWRIganderson** [comment]
@calbaker This may be a bit clunky....  I don't like the JSON approach because it involves converting data to text and back which is ridiculously inefficient.  CBOR could mitigate this.  The stackoverflow post below has a pretty easy solution to get all the paths into a list.  This may not be the solution, but I hate saying something won't work without a suggestion of an alternative.  

I was able to take the get_paths function with the two lines of code below the link to generate a list of paths.  The paths were strings.  In VS Code this all printed out on one line.  Spyder may look a bit different.  This list of paths would be super helpful for pulling out all the history vectors.  I think we could do some slick list comprehension to get all paths with history in the name to make it into a table.   

https://stackoverflow.com/questions/67352113/get-list-of-json-paths-in-python 

x=json.loads(bel.to_json())
paths = ['obj'+s for s in get_paths(x)]

The other thing that is worth considering is aggregating loco history data to summarize the consist.  I did some of this in the taconite case study.  Its not real flexible, but it would be a starting point to do something like this.  Having a sum of consist effort would be nice.  Also having an average SOC for all BELS in the consist would be cool.  One thing to consider though is what happens when we have BELs of different capacities.  The average SOC would need to be based on consist battery capacity and not a straight up average of SOC.  I hope this helps.  Please let me know what I can do to help support this effort.  This seems like a good feature to have.  Not all of us are Spyder lovers.

**Pull request closed without merging**